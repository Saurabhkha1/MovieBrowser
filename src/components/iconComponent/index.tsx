import React from 'react';
import {FontAwesomeIcon} from '@fortawesome/react-native-fontawesome';
import {IconProp} from '@fortawesome/fontawesome-svg-core';

type Props = {
  icon: IconProp;
  size?: number;
  color?: string;
};

export const IconComponent = ({icon, size, color}: Props) => {
  return <FontAwesomeIcon icon={icon} size={size} color={color} />;
};
