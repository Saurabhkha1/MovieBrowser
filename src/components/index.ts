export * from './iconComponent';
export * from './header';
export * from './movieComponent';
export * from './loader';
