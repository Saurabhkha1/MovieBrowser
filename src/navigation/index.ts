export * from './bottomTabs';
export * from './stack';
