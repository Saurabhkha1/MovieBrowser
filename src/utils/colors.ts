export const Colors = {
  black: '#000',
  white: '#fff',
  lightBlue: '#7DF9FF',
  darkBlue: '#00008b',
  brightBlue: '#0096FF',
  grey: '#ccc',
  darkGrey: '#5A5A5A',
  red: '#ff0000',
};
