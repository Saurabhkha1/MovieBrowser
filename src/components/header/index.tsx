import React, {useState, useRef, useEffect} from 'react';
import {
  View,
  Text,
  StyleSheet,
  TouchableOpacity,
  Animated,
  TextInput,
  Keyboard,
} from 'react-native';
import {moderateScale} from '../../utils/helper';
import {Colors} from '../../utils/colors';
import {IconComponent} from '../iconComponent';
import {IconProp} from '@fortawesome/fontawesome-svg-core';
import {Icons} from '../../utils/icons';

type Props = {
  title: string;
  icon: IconProp;
  onChangeText: (text: string) => void;
  searchValue: string;
};

export const Header = ({title, icon, onChangeText, searchValue}: Props) => {
  const [isInputVisible, setInputVisible] = useState(false);
  const slideAnim = useRef(new Animated.Value(-100)).current;
  const inputRef = useRef<TextInput>(null);

  useEffect(() => {
    Animated.timing(slideAnim, {
      toValue: isInputVisible ? 0 : -100,
      duration: 300,
      useNativeDriver: true,
    }).start(() =>
      isInputVisible ? inputRef?.current?.focus() : Keyboard.dismiss(),
    );
  }, [isInputVisible, slideAnim]);

  const toggleInputVisibility = () => {
    setInputVisible(!isInputVisible);
    onChangeText?.('');
  };

  return (
    <View style={styles.container}>
      <View style={styles.leftContainer}>
        <IconComponent
          icon={icon}
          size={moderateScale(25)}
          color={Colors.white}
        />
        <Text style={styles.heading}>{title}</Text>
      </View>
      <TouchableOpacity onPress={toggleInputVisibility}>
        <IconComponent
          icon={Icons.search}
          size={moderateScale(25)}
          color={Colors.white}
        />
      </TouchableOpacity>

      <Animated.View
        style={[styles.inputContainer, {transform: [{translateY: slideAnim}]}]}>
        <TextInput
          ref={inputRef}
          style={styles.input}
          value={searchValue}
          placeholder="Search for movies..."
          placeholderTextColor={Colors.grey}
          onChangeText={onChangeText}
        />
        <TouchableOpacity
          style={styles.closeInputButton}
          onPress={toggleInputVisibility}>
          <IconComponent icon={Icons.close} size={moderateScale(25)} />
        </TouchableOpacity>
      </Animated.View>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    padding: moderateScale(10),
    flexDirection: 'row',
    alignItems: 'center',
    backgroundColor: Colors.darkBlue,
    justifyContent: 'space-between',
    position: 'relative',
  },
  leftContainer: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  heading: {
    color: Colors.white,
    fontSize: moderateScale(20),
    fontWeight: '500',
    marginLeft: moderateScale(10),
  },
  inputContainer: {
    position: 'absolute',
    bottom: 0,
    left: 0,
    right: 0,
    backgroundColor: Colors.white,
    padding: moderateScale(5),
    paddingHorizontal: moderateScale(10),
    borderTopWidth: moderateScale(1),
    borderTopColor: Colors.grey,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
  },
  input: {
    height: moderateScale(40),
    borderColor: Colors.grey,
    borderWidth: moderateScale(1),
    borderRadius: moderateScale(5),
    paddingHorizontal: moderateScale(10),
    color: Colors.black,
    width: '90%',
  },
  closeInputButton: {
    width: '10%',
    alignItems: 'center',
    justifyContent: 'center',
  },
});
