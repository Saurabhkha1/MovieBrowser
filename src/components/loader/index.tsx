import {View, ActivityIndicator, StyleSheet} from 'react-native';
import React from 'react';
import {Colors} from '../../utils/colors';

type Props = {
  size?: 'full' | 'normal';
  color?: string;
};

export const Loader = ({size, color = Colors.black}: Props) => {
  return (
    <View style={size === 'full' ? styles.fullScreen : styles.normal}>
      <ActivityIndicator size={'large'} color={color} />
    </View>
  );
};

const styles = StyleSheet.create({
  fullScreen: {
    height: '100%',
    alignItems: 'center',
    justifyContent: 'center',
  },
  normal: {
    alignItems: 'center',
    justifyContent: 'center',
  },
});
