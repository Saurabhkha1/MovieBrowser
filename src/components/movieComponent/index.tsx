import {
  View,
  Text,
  StyleSheet,
  TouchableOpacity,
  Image,
  Platform,
} from 'react-native';
import React from 'react';
import {deviceWidth, moderateScale} from '../../utils/helper';
import {Colors} from '../../utils/colors';
import {Movie} from '../../redux/movieSlice';
import {IMAGE_BASE_URL} from '../../utils/constants';
import {useNavigation} from '@react-navigation/native';
import {NativeStackNavigationProp} from '@react-navigation/native-stack';
import {RootStackParamList} from '../../navigation';

type Props = {
  movie: Movie;
};

export const MovieComponent = ({movie}: Props) => {
  const navigation =
    useNavigation<NativeStackNavigationProp<RootStackParamList>>();
  return (
    <TouchableOpacity
      style={styles.container}
      onPress={() => navigation.navigate('MovieDetails', {movie: movie})}>
      <Image
        source={{
          uri: IMAGE_BASE_URL + movie?.poster_path,
        }}
        style={styles.poster}
      />
      <View style={styles.details}>
        <Text style={styles.name} numberOfLines={2} ellipsizeMode="tail">
          {movie?.title}
        </Text>
        <Text style={styles.dateWithNoMargin}>Release Date:</Text>
        <Text style={styles.date}>{movie?.release_date}</Text>
        <Text style={styles.date}>
          Rating: {movie?.vote_average?.toFixed(1)}/10
        </Text>
      </View>
    </TouchableOpacity>
  );
};

const styles = StyleSheet.create({
  container: {
    width: (deviceWidth - moderateScale(30)) / 2,
    height: Platform.OS === 'android' ? moderateScale(450) : moderateScale(400),
    borderRadius: moderateScale(10),
    backgroundColor: Colors.lightBlue,
    marginRight: moderateScale(10),
    marginBottom: moderateScale(10),
  },
  poster: {
    width: '100%',
    height: '70%',
    borderRadius: moderateScale(10),
  },
  details: {
    padding: moderateScale(5),
  },
  name: {
    color: Colors.black,
    fontSize: moderateScale(15),
    fontWeight: '500',
    marginBottom: moderateScale(10),
  },
  date: {
    color: Colors.black,
    fontSize: moderateScale(15),
    marginBottom: moderateScale(10),
  },
  dateWithNoMargin: {
    color: Colors.black,
    fontSize: moderateScale(15),
  },
});
