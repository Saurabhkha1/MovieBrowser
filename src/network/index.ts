import axios from 'axios';
import {BASE_URL, BEARER_TOKEN} from '../utils/constants';

type Movie = {
  id: number;
  title: string;
  poster_path: string;
  release_date: string;
  vote_average: number;
  overview: string;
};

export const fetchMovies = async (
  endpoint: string,
  page: number = 1,
): Promise<Movie[]> => {
  const response = await axios.get(`${BASE_URL}/movie/${endpoint}`, {
    params: {language: 'en-US', page},
    headers: {
      Authorization: `Bearer ${BEARER_TOKEN}`,
      Accept: 'application/json',
    },
  });
  return response.data.results;
};
