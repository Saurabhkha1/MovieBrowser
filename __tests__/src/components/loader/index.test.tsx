import React from 'react';
import {create, act, ReactTestRenderer} from 'react-test-renderer';
import {View, ActivityIndicator} from 'react-native';
import {Loader} from '../../../../src/components';
import {Colors} from '../../../../src/utils/colors';

describe('Loader Component', () => {
  it('renders correctly with default props', () => {
    let component: ReactTestRenderer | undefined;
    act(() => {
      component = create(<Loader />);
    });

    const tree = component?.toJSON();
    expect(tree).toMatchSnapshot();

    const view = component?.root.findByType(View);
    const activityIndicator = component?.root.findByType(ActivityIndicator);

    expect(view?.props.style).toEqual(
      expect.objectContaining({
        alignItems: 'center',
        justifyContent: 'center',
      }),
    );

    expect(activityIndicator?.props.size).toBe('large');
    expect(activityIndicator?.props.color).toBe(Colors.black);
  });

  it('renders correctly with full size', () => {
    let component: ReactTestRenderer | undefined;
    act(() => {
      component = create(<Loader size="full" />);
    });

    const tree = component?.toJSON();
    expect(tree).toMatchSnapshot();

    const view = component?.root.findByType(View);

    expect(view?.props.style).toEqual(
      expect.objectContaining({
        height: '100%',
        alignItems: 'center',
        justifyContent: 'center',
      }),
    );
  });

  it('renders correctly with custom color', () => {
    const customColor = 'red';
    let component: ReactTestRenderer | undefined;
    act(() => {
      component = create(<Loader color={customColor} />);
    });

    const tree = component?.toJSON();
    expect(tree).toMatchSnapshot();

    const activityIndicator = component?.root.findByType(ActivityIndicator);

    expect(activityIndicator?.props.color).toBe(customColor);
  });

  it('renders correctly with normal size and custom color', () => {
    const customColor = 'blue';
    let component: ReactTestRenderer | undefined;
    act(() => {
      component = create(<Loader size="normal" color={customColor} />);
    });

    const tree = component?.toJSON();
    expect(tree).toMatchSnapshot();

    const view = component?.root.findByType(View);
    const activityIndicator = component?.root.findByType(ActivityIndicator);

    expect(view?.props.style).toEqual(
      expect.objectContaining({
        alignItems: 'center',
        justifyContent: 'center',
      }),
    );

    expect(activityIndicator?.props.size).toBe('large');
    expect(activityIndicator?.props.color).toBe(customColor);
  });
});
