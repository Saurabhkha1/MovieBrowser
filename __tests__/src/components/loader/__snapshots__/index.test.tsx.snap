// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loader Component renders correctly with custom color 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "justifyContent": "center",
    }
  }
>
  <ActivityIndicator
    color="red"
    size="large"
  />
</View>
`;

exports[`Loader Component renders correctly with default props 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "justifyContent": "center",
    }
  }
>
  <ActivityIndicator
    color="#000"
    size="large"
  />
</View>
`;

exports[`Loader Component renders correctly with full size 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "height": "100%",
      "justifyContent": "center",
    }
  }
>
  <ActivityIndicator
    color="#000"
    size="large"
  />
</View>
`;

exports[`Loader Component renders correctly with normal size and custom color 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "justifyContent": "center",
    }
  }
>
  <ActivityIndicator
    color="blue"
    size="large"
  />
</View>
`;
